File: python/packages/handoffkit/handoffkit/csp/distributed.py
"""Initial deterministic worker registry and scheduler for HK-CSP 1.18."""

from __future__ import annotations

import threading
import time
from collections import OrderedDict, deque
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from enum import Enum
from typing import Any
from uuid import uuid4

from handoffkit.csp.models import (
    DistributedJob,
    JobAssignment,
    WorkerCapabilities,
    WorkerHeartbeat,
    utc_now,
)


class WorkerStatus(str, Enum):
    ONLINE = "online"
    SUSPECT = "suspect"
    OFFLINE = "offline"


@dataclass(frozen=True)
class WorkerRecord:
    capabilities: WorkerCapabilities
    status: WorkerStatus
    heartbeat_sequence: int
    active_jobs: int
    load: float
    last_seen: float
    metadata: dict[str, Any] = field(default_factory=dict)

    @property
    def worker_id(self) -> str:
        return self.capabilities.worker_id


class WorkerRegistry:
    """Thread-safe in-memory registry with heartbeat expiry and routing."""

    def __init__(
        self,
        *,
        suspect_after: float = 15.0,
        offline_after: float = 45.0,
        clock: Any = time.monotonic,
    ) -> None:
        if suspect_after <= 0 or offline_after < suspect_after:
            raise ValueError("heartbeat thresholds are invalid")
        self.suspect_after = suspect_after
        self.offline_after = offline_after
        self._clock = clock
        self._records: dict[str, WorkerRecord] = {}
        self._lock = threading.RLock()

    def register(self, capabilities: WorkerCapabilities) -> WorkerRecord:
        record = WorkerRecord(
            capabilities=capabilities,
            status=WorkerStatus.ONLINE,
            heartbeat_sequence=0,
            active_jobs=0,
            load=0.0,
            last_seen=float(self._clock()),
        )
        with self._lock:
            self._records[capabilities.worker_id] = record
        return record

    def heartbeat(self, heartbeat: WorkerHeartbeat) -> bool:
        with self._lock:
            current = self._records.get(heartbeat.worker_id)
            if current is None:
                raise KeyError(f"unknown worker {heartbeat.worker_id!r}")
            if heartbeat.sequence <= current.heartbeat_sequence:
                return False
            self._records[heartbeat.worker_id] = WorkerRecord(
                capabilities=current.capabilities,
                status=WorkerStatus.ONLINE,
                heartbeat_sequence=heartbeat.sequence,
                active_jobs=heartbeat.active_jobs,
                load=heartbeat.load,
                last_seen=float(self._clock()),
                metadata=dict(heartbeat.metadata),
            )
            return True

    def mark_disconnected(self, worker_id: str) -> None:
        with self._lock:
            current = self._required(worker_id)
            self._records[worker_id] = self._replace(
                current,
                status=WorkerStatus.OFFLINE,
            )

    def expire(self) -> list[str]:
        now = float(self._clock())
        changed: list[str] = []
        with self._lock:
            for worker_id, current in list(self._records.items()):
                age = now - current.last_seen
                status = (
                    WorkerStatus.OFFLINE
                    if age >= self.offline_after
                    else WorkerStatus.SUSPECT
                    if age >= self.suspect_after
                    else WorkerStatus.ONLINE
                )
                if status is not current.status:
                    self._records[worker_id] = self._replace(current, status=status)
                    changed.append(worker_id)
        return changed

    def reserve(self, required_capabilities: list[str]) -> WorkerRecord | None:
        required = set(required_capabilities)
        with self._lock:
            candidates = [
                record
                for record in self._records.values()
                if record.status is WorkerStatus.ONLINE
                and required.issubset(set(record.capabilities.operations))
            ]
            if not candidates:
                return None
            chosen = min(
                candidates,
                key=lambda item: (item.load, item.active_jobs, item.worker_id),
            )
            reserved = self._replace(chosen, active_jobs=chosen.active_jobs + 1)
            self._records[chosen.worker_id] = reserved
            return reserved

    def release(self, worker_id: str) -> None:
        with self._lock:
            current = self._required(worker_id)
            self._records[worker_id] = self._replace(
                current,
                active_jobs=max(0, current.active_jobs - 1),
            )

    def get(self, worker_id: str) -> WorkerRecord | None:
        with self._lock:
            return self._records.get(worker_id)

    def list(self) -> list[WorkerRecord]:
        with self._lock:
            return sorted(self._records.values(), key=lambda item: item.worker_id)

    def _required(self, worker_id: str) -> WorkerRecord:
        record = self._records.get(worker_id)
        if record is None:
            raise KeyError(f"unknown worker {worker_id!r}")
        return record

    @staticmethod
    def _replace(record: WorkerRecord, **changes: Any) -> WorkerRecord:
        values = {
            "capabilities": record.capabilities,
            "status": record.status,
            "heartbeat_sequence": record.heartbeat_sequence,
            "active_jobs": record.active_jobs,
            "load": record.load,
            "last_seen": record.last_seen,
            "metadata": dict(record.metadata),
        }
        values.update(changes)
        return WorkerRecord(**values)


@dataclass(frozen=True)
class SchedulerSnapshot:
    queued: int
    assigned: int
    completed: int
    failed: int
    seen_jobs: int


@dataclass
class _AssignmentState:
    assignment: JobAssignment
    job: DistributedJob


class DistributedScheduler:
    """Bounded, lease-based initial scheduler; not a global cluster scheduler."""

    def __init__(
        self,
        registry: WorkerRegistry,
        *,
        max_attempts: int = 3,
        lease_seconds: float = 30.0,
        queue_capacity: int = 4096,
        dedup_capacity: int = 100_000,
    ) -> None:
        if max_attempts < 1 or lease_seconds <= 0:
            raise ValueError("scheduler retry and lease limits must be positive")
        if queue_capacity < 1 or dedup_capacity < 1:
            raise ValueError("scheduler capacities must be positive")
        self.registry = registry
        self.max_attempts = max_attempts
        self.lease_seconds = lease_seconds
        self.queue_capacity = queue_capacity
        self.dedup_capacity = dedup_capacity
        self._queue: deque[DistributedJob] = deque()
        self._assigned: dict[str, _AssignmentState] = {}
        self._attempts: dict[str, int] = {}
        self._seen: OrderedDict[str, str] = OrderedDict()
        self._completed = 0
        self._failed = 0
        self._lock = threading.RLock()

    def submit(self, job: DistributedJob) -> bool:
        with self._lock:
            if job.idempotency_key in self._seen:
                return False
            if len(self._queue) >= self.queue_capacity:
                raise RuntimeError("distributed scheduler queue is at capacity")
            self._seen[job.idempotency_key] = job.job_id
            while len(self._seen) > self.dedup_capacity:
                self._seen.popitem(last=False)
            self._attempts.setdefault(job.job_id, 0)
            self._queue.append(job)
            return True

    def schedule(self) -> JobAssignment | None:
        with self._lock:
            for _ in range(len(self._queue)):
                job = self._queue.popleft()
                worker = self.registry.reserve(job.requested_capabilities)
                if worker is None:
                    self._queue.append(job)
                    continue
                attempt = self._attempts.get(job.job_id, 0) + 1
                self._attempts[job.job_id] = attempt
                assigned = datetime.now(timezone.utc)
                assignment = JobAssignment(
                    assignment_id=f"assignment-{uuid4().hex}",
                    job_id=job.job_id,
                    worker_id=worker.worker_id,
                    attempt=attempt,
                    assigned_at=assigned.isoformat().replace("+00:00", "Z"),
                    lease_deadline=(assigned + timedelta(seconds=self.lease_seconds))
                    .isoformat()
                    .replace("+00:00", "Z"),
                    payload=job.payload,
                    metadata={
                        **job.metadata,
                        "operation": job.operation,
                        "idempotency_key": job.idempotency_key,
                    },
                )
                self._assigned[assignment.assignment_id] = _AssignmentState(
                    assignment,
                    job,
                )
                return assignment
            return None

    def complete(self, assignment_id: str) -> bool:
        with self._lock:
            state = self._assigned.pop(assignment_id, None)
            if state is None:
                return False
            self.registry.release(state.assignment.worker_id)
            self._completed += 1
            return True

    def fail(self, assignment_id: str, *, retryable: bool = True) -> bool:
        with self._lock:
            state = self._assigned.pop(assignment_id, None)
            if state is None:
                return False
            self.registry.release(state.assignment.worker_id)
            if (
                retryable
                and state.assignment.attempt < self.max_attempts
                and len(self._queue) < self.queue_capacity
            ):
                self._queue.appendleft(state.job)
            else:
                self._failed += 1
            return True

    def recover_worker(self, worker_id: str) -> int:
        with self._lock:
            assignment_ids = [
                assignment_id
                for assignment_id, state in self._assigned.items()
                if state.assignment.worker_id == worker_id
            ]
            for assignment_id in assignment_ids:
                self.fail(assignment_id, retryable=True)
            return len(assignment_ids)

    def recover_expired(self, *, now: datetime | None = None) -> int:
        current = now or datetime.now(timezone.utc)
        with self._lock:
            expired = [
                assignment_id
                for assignment_id, state in self._assigned.items()
                if datetime.fromisoformat(state.assignment.lease_deadline.replace("Z", "+00:00"))
                <= current
            ]
            for assignment_id in expired:
                self.fail(assignment_id, retryable=True)
            return len(expired)

    def snapshot(self) -> SchedulerSnapshot:
        with self._lock:
            return SchedulerSnapshot(
                queued=len(self._queue),
                assigned=len(self._assigned),
                completed=self._completed,
                failed=self._failed,
                seen_jobs=len(self._seen),
            )


def heartbeat_now(
    worker_id: str,
    *,
    sequence: int,
    active_jobs: int,
    load: float,
    metadata: dict[str, Any] | None = None,
) -> WorkerHeartbeat:
    return WorkerHeartbeat(
        worker_id=worker_id,
        sequence=sequence,
        active_jobs=active_jobs,
        load=load,
        timestamp=utc_now(),
        metadata=dict(metadata or {}),
    )

File: js/packages/csp/src/distributed.js
import {
  DistributedJob,
  JobAssignment,
  WorkerCapabilities,
  WorkerHeartbeat,
} from "./contracts.js";
import { SecurityError } from "./security.js";

export const WorkerStatus = Object.freeze({ ONLINE: "online", SUSPECT: "suspect", OFFLINE: "offline" });
export const SCHEDULER_STATE_FORMAT = "handoffkit.scheduler.state";
export const SCHEDULER_STATE_FORMAT_VERSION = 1;
const MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER;

function migrateSchedulerState(value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new TypeError("scheduler state root is invalid");
  if (value.format !== SCHEDULER_STATE_FORMAT) return { value, migrated: false };
  if (value.format_version === SCHEDULER_STATE_FORMAT_VERSION) return { value, migrated: false };
  if (value.format_version !== 0) throw new TypeError("scheduler state format is unsupported");
  const expected = ["completed", "failed", "format", "format_version", "generation", "inflight", "queued", "seen"];
  if (Object.keys(value).sort().join("\0") !== expected.sort().join("\0")) throw new TypeError("scheduler v0 state fields are invalid");
  return {
    value: { ...value, format_version: SCHEDULER_STATE_FORMAT_VERSION, interrupted: [] },
    migrated: true,
  };
}

function schedulerCommitApplied(error) {
  return error?.committed === true || error?.details?.committed === true;
}

function identifier(prefix) {
  return `${prefix}-${globalThis.crypto?.randomUUID?.() ?? `${Date.now()}-${Math.random().toString(16).slice(2)}`}`;
}

export class WorkerRecord {
  constructor({ capabilities, status = WorkerStatus.ONLINE, heartbeatSequence = 0, activeJobs = 0, load = 0, lastSeen = Date.now(), metadata = {} }) {
    this.capabilities = capabilities instanceof WorkerCapabilities ? capabilities : new WorkerCapabilities(capabilities);
    this.status = status;
    this.heartbeatSequence = heartbeatSequence;
    this.activeJobs = activeJobs;
    this.load = load;
    this.lastSeen = lastSeen;
    this.metadata = { ...metadata };
  }
  get workerId() { return this.capabilities.toWire().worker_id; }
  with(changes) {
    return new WorkerRecord({
      capabilities: this.capabilities,
      status: this.status,
      heartbeatSequence: this.heartbeatSequence,
      activeJobs: this.activeJobs,
      load: this.load,
      lastSeen: this.lastSeen,
      metadata: this.metadata,
      ...changes,
    });
  }
}

export class WorkerRegistry {
  constructor({ suspectAfterMs = 15_000, offlineAfterMs = 45_000, clock = Date.now } = {}) {
    if (suspectAfterMs <= 0 || offlineAfterMs < suspectAfterMs) throw new TypeError("heartbeat thresholds are invalid");
    this.suspectAfterMs = suspectAfterMs;
    this.offlineAfterMs = offlineAfterMs;
    this.clock = clock;
    this.records = new Map();
  }
  register(value) {
    const capabilities = value instanceof WorkerCapabilities ? value : new WorkerCapabilities(value);
    const record = new WorkerRecord({ capabilities, lastSeen: this.clock() });
    this.records.set(record.workerId, record);
    return record;
  }
  heartbeat(value) {
    const heartbeat = value instanceof WorkerHeartbeat ? value : new WorkerHeartbeat(value);
    const wire = heartbeat.toWire();
    const current = this.required(wire.worker_id);
    if (wire.sequence <= current.heartbeatSequence) return false;
    this.records.set(wire.worker_id, current.with({
      status: WorkerStatus.ONLINE,
      heartbeatSequence: wire.sequence,
      activeJobs: wire.active_jobs,
      load: wire.load,
      lastSeen: this.clock(),
      metadata: wire.metadata,
    }));
    return true;
  }
  markDisconnected(workerId) { const current = this.required(workerId); this.records.set(workerId, current.with({ status: WorkerStatus.OFFLINE })); }
  expire() {
    const now = this.clock();
    const changed = [];
    for (const [workerId, record] of this.records) {
      const age = now - record.lastSeen;
      const status = age >= this.offlineAfterMs ? WorkerStatus.OFFLINE : age >= this.suspectAfterMs ? WorkerStatus.SUSPECT : WorkerStatus.ONLINE;
      if (status !== record.status) { this.records.set(workerId, record.with({ status })); changed.push(workerId); }
    }
    return changed;
  }
  reserve(requiredCapabilities = []) {
    const required = new Set(requiredCapabilities);
    const candidates = [...this.records.values()].filter((record) => {
      const operations = new Set(record.capabilities.toWire().operations ?? []);
      return record.status === WorkerStatus.ONLINE && [...required].every((item) => operations.has(item));
    });
    candidates.sort((left, right) => left.load - right.load || left.activeJobs - right.activeJobs || left.workerId.localeCompare(right.workerId));
    const chosen = candidates[0];
    if (!chosen) return null;
    const reserved = chosen.with({ activeJobs: chosen.activeJobs + 1 });
    this.records.set(chosen.workerId, reserved);
    return reserved;
  }
  release(workerId) { const current = this.required(workerId); this.records.set(workerId, current.with({ activeJobs: Math.max(0, current.activeJobs - 1) })); }
  get(workerId) { return this.records.get(workerId) ?? null; }
  list() { return [...this.records.values()].sort((left, right) => left.workerId.localeCompare(right.workerId)); }
  required(workerId) { const record = this.records.get(workerId); if (!record) throw new TypeError(`unknown worker ${workerId}`); return record; }
}

export class DistributedScheduler {
  constructor(registry, {
    maxAttempts = 3,
    leaseMs = 30_000,
    queueCapacity = 4096,
    dedupCapacity = 100_000,
    stateStore = null,
    autoResume = false,
  } = {}) {
    if (!(registry instanceof WorkerRegistry)) throw new TypeError("registry must be a WorkerRegistry");
    if (![maxAttempts, leaseMs, queueCapacity, dedupCapacity].every(Number.isSafeInteger)
      || maxAttempts < 1 || leaseMs < 1 || queueCapacity < 1 || dedupCapacity < 1) {
      throw new TypeError("scheduler limits must be positive safe integers");
    }
    if (typeof autoResume !== "boolean") throw new TypeError("autoResume must be a boolean");
    if (stateStore) {
      for (const name of ["load", "commit", "quarantine"]) {
        if (typeof stateStore[name] !== "function") throw new TypeError(`stateStore must implement ${name}()`);
      }
    }
    Object.assign(this, { registry, maxAttempts, leaseMs, queueCapacity, dedupCapacity, stateStore, autoResume });
    this.queue = [];
    this.assignments = new Map();
    this.interrupted = new Map();
    this.seen = new Map();
    this.completed = 0;
    this.failed = 0;
    this.generation = 0;
    if (stateStore) this.loadState();
    if (stateStore && autoResume) this.autoResumeInterrupted();
  }
  submit(value) {
    const job = value instanceof DistributedJob ? value : new DistributedJob(value);
    const wire = job.toWire();
    if (wire.metadata && Object.prototype.hasOwnProperty.call(wire.metadata, "require_exactly_once")
      && wire.metadata.require_exactly_once !== false) {
      throw new SecurityError(
        "Exactly-once external effects are unavailable; refusing fallback to at-least-once.",
        {
          code: "exactly_once_unavailable",
          details: { runtime: "javascript" },
        },
      );
    }
    if (this.isDuplicate(wire)) return false;
    if (this.queue.length >= this.queueCapacity) throw new Error("distributed scheduler queue is at capacity");
    const previous = this.captureState();
    try {
      this.claimSeen(wire);
      this.queue.push({ job, attempt: 1 });
      this.persist();
    } catch (error) {
      if (!schedulerCommitApplied(error)) this.restoreState(previous);
      throw error;
    }
    return true;
  }
  schedule() {
    if (this.assignments.size + this.interrupted.size >= this.queueCapacity) return null;
    const queued = this.queue.length;
    for (let index = 0; index < queued; index += 1) {
      const previous = this.captureState();
      const queuedState = this.queue.shift();
      const { job } = queuedState;
      const wire = job.toWire();
      const worker = this.registry.reserve(wire.requested_capabilities);
      if (!worker) { this.queue.push(queuedState); continue; }
      const assigned = Date.now();
      const assignment = new JobAssignment({
        assignment_id: identifier("assignment"),
        job_id: wire.job_id,
        worker_id: worker.workerId,
        attempt: queuedState.attempt,
        assigned_at: new Date(assigned).toISOString(),
        lease_deadline: new Date(assigned + this.leaseMs).toISOString(),
        payload: wire.payload,
        metadata: { ...wire.metadata, operation: wire.operation, idempotency_key: wire.idempotency_key },
      });
      this.assignments.set(assignment.toWire().assignment_id, { assignment, job });
      try {
        this.persist();
      } catch (error) {
        if (!schedulerCommitApplied(error)) {
          this.restoreState(previous);
          this.registry.release(worker.workerId);
        }
        throw error;
      }
      return assignment;
    }
    return null;
  }
  complete(assignmentId) {
    const state = this.assignments.get(assignmentId);
    if (!state) return false;
    const previous = this.captureState();
    try {
      this.assignments.delete(assignmentId);
      this.completed += 1;
      this.persist();
    } catch (error) {
      if (schedulerCommitApplied(error)) this.registry.release(state.assignment.toWire().worker_id);
      else this.restoreState(previous);
      throw error;
    }
    this.registry.release(state.assignment.toWire().worker_id);
    return true;
  }
  fail(assignmentId, { retryable = true } = {}) {
    const state = this.assignments.get(assignmentId);
    if (!state) return false;
    const previous = this.captureState();
    try {
      this.failState(assignmentId, retryable);
      this.persist();
    } catch (error) {
      if (schedulerCommitApplied(error)) this.registry.release(state.assignment.toWire().worker_id);
      else this.restoreState(previous);
      throw error;
    }
    this.registry.release(state.assignment.toWire().worker_id);
    return true;
  }
  recoverWorker(workerId) {
    const states = [...this.assignments]
      .filter(([, state]) => state.assignment.toWire().worker_id === workerId);
    if (states.length === 0) return 0;
    const previous = this.captureState();
    try {
      for (const [assignmentId] of states) this.failState(assignmentId, true);
      this.persist();
    } catch (error) {
      if (schedulerCommitApplied(error)) {
        for (const [, state] of states) this.registry.release(state.assignment.toWire().worker_id);
      } else this.restoreState(previous);
      throw error;
    }
    for (const [, state] of states) this.registry.release(state.assignment.toWire().worker_id);
    return states.length;
  }
  recoverExpired({ now = Date.now() } = {}) {
    const states = [...this.assignments]
      .filter(([, state]) => Date.parse(state.assignment.toWire().lease_deadline) <= now);
    if (states.length === 0) return 0;
    const previous = this.captureState();
    try {
      for (const [assignmentId] of states) this.failState(assignmentId, true);
      this.persist();
    } catch (error) {
      if (schedulerCommitApplied(error)) {
        for (const [, state] of states) this.registry.release(state.assignment.toWire().worker_id);
      } else this.restoreState(previous);
      throw error;
    }
    for (const [, state] of states) this.registry.release(state.assignment.toWire().worker_id);
    return states.length;
  }
  listInterrupted() {
    return [...this.interrupted]
      .sort(([left], [right]) => left.localeCompare(right))
      .map(([, state]) => state.assignment);
  }
  retryInterrupted(assignmentId) {
    const state = this.interrupted.get(assignmentId);
    if (!state) return false;
    if (this.queue.length >= this.queueCapacity) throw new Error("distributed scheduler queue is at capacity");
    const previous = this.captureState();
    try {
      this.interrupted.delete(assignmentId);
      const nextAttempt = state.assignment.toWire().attempt + 1;
      if (nextAttempt <= this.maxAttempts) this.queue.unshift({ job: state.job, attempt: nextAttempt });
      else this.failed += 1;
      this.persist();
    } catch (error) {
      if (!schedulerCommitApplied(error)) this.restoreState(previous);
      throw error;
    }
    return true;
  }
  autoResumeInterrupted() {
    for (const assignmentId of [...this.interrupted.keys()].sort()) this.retryInterrupted(assignmentId);
  }
  failInterrupted(assignmentId) {
    if (!this.interrupted.has(assignmentId)) return false;
    const previous = this.captureState();
    try {
      this.interrupted.delete(assignmentId);
      this.failed += 1;
      this.persist();
    } catch (error) {
      if (!schedulerCommitApplied(error)) this.restoreState(previous);
      throw error;
    }
    return true;
  }
  snapshot() {
    return {
      queued: this.queue.length,
      assigned: this.assignments.size,
      interrupted: this.interrupted.size,
      completed: this.completed,
      failed: this.failed,
      seen_jobs: this.seen.size,
    };
  }
  get stateGeneration() { return this.generation; }
  failState(assignmentId, retryable) {
    const state = this.assignments.get(assignmentId);
    this.assignments.delete(assignmentId);
    const nextAttempt = state.assignment.toWire().attempt + 1;
    if (retryable && nextAttempt <= this.maxAttempts && this.queue.length < this.queueCapacity) {
      this.queue.unshift({ job: state.job, attempt: nextAttempt });
    } else {
      this.failed += 1;
    }
  }
  activeJobIds() {
    return new Set([
      ...this.queue.map((state) => state.job.toWire().job_id),
      ...[...this.assignments.values()].map((state) => state.job.toWire().job_id),
      ...[...this.interrupted.values()].map((state) => state.job.toWire().job_id),
    ]);
  }
  isDuplicate(wire) {
    return this.seen.has(wire.idempotency_key)
      || [...this.seen.values()].includes(wire.job_id)
      || this.activeJobIds().has(wire.job_id);
  }
  claimSeen(wire) {
    const activeJobIds = this.activeJobIds();
    while (this.seen.size >= this.dedupCapacity) {
      const evictable = [...this.seen].find(([, jobId]) => !activeJobIds.has(jobId));
      if (!evictable) throw new Error("distributed scheduler deduplication state is at capacity");
      this.seen.delete(evictable[0]);
    }
    this.seen.set(wire.idempotency_key, wire.job_id);
  }
  captureState() {
    return {
      queue: [...this.queue],
      assignments: new Map(this.assignments),
      interrupted: new Map(this.interrupted),
      seen: new Map(this.seen),
      completed: this.completed,
      failed: this.failed,
      generation: this.generation,
    };
  }
  restoreState(state) {
    Object.assign(this, state);
  }
  statePayload(generation) {
    if (![this.completed, this.failed, generation].every((value) => Number.isSafeInteger(value) && value >= 0)) {
      throw new TypeError("scheduler state counters exceed the interoperable integer range");
    }
    return {
      completed: this.completed,
      failed: this.failed,
      format: SCHEDULER_STATE_FORMAT,
      format_version: SCHEDULER_STATE_FORMAT_VERSION,
      generation,
      inflight: [...this.assignments]
        .sort(([left], [right]) => left.localeCompare(right))
        .map(([, state]) => ({ assignment: state.assignment.toWire(), job: state.job.toWire() })),
      interrupted: [...this.interrupted]
        .sort(([left], [right]) => left.localeCompare(right))
        .map(([, state]) => ({
          assignment: state.assignment.toWire(),
          job: state.job.toWire(),
          reason: state.reason,
        })),
      queued: this.queue.map((state) => ({ attempt: state.attempt, job: state.job.toWire() })),
      seen: [...this.seen].map(([idempotencyKey, jobId]) => ({
        idempotency_key: idempotencyKey,
        job_id: jobId,
      })),
    };
  }
  persist() {
    if (!this.stateStore) return;
    const generation = this.generation + 1;
    try {
      this.stateStore.commit(this.statePayload(generation));
    } catch (error) {
      if (schedulerCommitApplied(error)) this.generation = generation;
      throw error;
    }
    this.generation = generation;
  }
  loadState() {
    const value = this.stateStore.load();
    if (value === null || value === undefined) return;
    try {
      const migrated = migrateSchedulerState(value);
      if (migrated.migrated) this.stateStore.commit(migrated.value);
      this.decodeState(migrated.value);
    } catch (error) {
      this.stateStore.quarantine(`invalid scheduler state: ${error?.constructor?.name ?? "Error"}`);
    }
    if (this.assignments.size > 0) {
      for (const [assignmentId, state] of this.assignments) {
        this.interrupted.set(assignmentId, { ...state, reason: "scheduler_restart" });
      }
      this.assignments.clear();
      this.persist();
    }
  }
  decodeState(value) {
    const required = ["completed", "failed", "format", "format_version", "generation", "inflight", "interrupted", "queued", "seen"];
    if (!value || typeof value !== "object" || Array.isArray(value)
      || Object.keys(value).sort().join("\0") !== [...required].sort().join("\0")) {
      throw new TypeError("scheduler state fields are invalid");
    }
    if (value.format !== SCHEDULER_STATE_FORMAT || value.format_version !== SCHEDULER_STATE_FORMAT_VERSION) {
      throw new TypeError("scheduler state format is unsupported");
    }
    for (const name of ["completed", "failed", "generation"]) {
      if (!Number.isSafeInteger(value[name]) || value[name] < 0) throw new TypeError(`scheduler state ${name} is invalid`);
    }
    for (const name of ["inflight", "interrupted", "queued", "seen"]) {
      if (!Array.isArray(value[name])) throw new TypeError("scheduler state collections are invalid");
    }
    if (value.queued.length > this.queueCapacity) throw new TypeError("scheduler queue exceeds configured capacity");
    if (value.inflight.length + value.interrupted.length > this.queueCapacity) throw new TypeError("scheduler interrupted state exceeds configured capacity");
    if (value.seen.length > this.dedupCapacity) throw new TypeError("scheduler dedup state exceeds configured capacity");

    const jobIds = new Set();
    const activeIdentities = new Map();
    const recordActive = (job) => {
      const wire = job.toWire();
      if (jobIds.has(wire.job_id) || activeIdentities.has(wire.idempotency_key)) {
        throw new TypeError("scheduler job is duplicated");
      }
      jobIds.add(wire.job_id);
      activeIdentities.set(wire.idempotency_key, wire.job_id);
    };
    const queue = value.queued.map((raw) => {
      if (!raw || typeof raw !== "object" || Object.keys(raw).sort().join("\0") !== "attempt\0job") throw new TypeError("queued scheduler record is invalid");
      if (!Number.isSafeInteger(raw.attempt) || raw.attempt < 1 || raw.attempt > this.maxAttempts) throw new TypeError("queued scheduler attempt is invalid");
      const job = new DistributedJob(raw.job);
      recordActive(job);
      return { job, attempt: raw.attempt };
    });
    const assignments = new Map();
    for (const raw of value.inflight) {
      const state = this.decodeAssignmentRecord(raw, false);
      const assignmentId = state.assignment.toWire().assignment_id;
      if (assignments.has(assignmentId)) throw new TypeError("scheduler assignment is duplicated");
      assignments.set(assignmentId, state);
      recordActive(state.job);
    }
    const interrupted = new Map();
    for (const raw of value.interrupted) {
      const state = this.decodeAssignmentRecord(raw, true);
      const assignmentId = state.assignment.toWire().assignment_id;
      if (assignments.has(assignmentId) || interrupted.has(assignmentId)) throw new TypeError("scheduler interrupted assignment is duplicated");
      interrupted.set(assignmentId, { ...state, reason: raw.reason });
      recordActive(state.job);
    }
    const seen = new Map();
    const seenJobIds = new Set();
    for (const raw of value.seen) {
      if (!raw || typeof raw !== "object" || Object.keys(raw).sort().join("\0") !== "idempotency_key\0job_id"
        || typeof raw.idempotency_key !== "string" || raw.idempotency_key.length === 0
        || typeof raw.job_id !== "string" || raw.job_id.length === 0
        || seen.has(raw.idempotency_key) || seenJobIds.has(raw.job_id)) {
        throw new TypeError("scheduler dedup record is invalid");
      }
      seen.set(raw.idempotency_key, raw.job_id);
      seenJobIds.add(raw.job_id);
    }
    for (const [key, jobId] of activeIdentities) {
      if (seen.get(key) !== jobId) throw new TypeError("scheduler active job is missing its dedup identity");
    }
    Object.assign(this, {
      queue,
      assignments,
      interrupted,
      seen,
      completed: value.completed,
      failed: value.failed,
      generation: value.generation,
    });
  }
  decodeAssignmentRecord(raw, interrupted) {
    const expected = interrupted ? "assignment\0job\0reason" : "assignment\0job";
    if (!raw || typeof raw !== "object" || Array.isArray(raw)
      || Object.keys(raw).sort().join("\0") !== expected) {
      throw new TypeError("scheduler assignment record is invalid");
    }
    if (interrupted && raw.reason !== "scheduler_restart") throw new TypeError("interrupted scheduler reason is invalid");
    const assignment = new JobAssignment(raw.assignment);
    const job = new DistributedJob(raw.job);
    if (assignment.toWire().job_id !== job.toWire().job_id) throw new TypeError("scheduler assignment job identity is inconsistent");
    return { assignment, job };
  }
}

export function heartbeatNow(workerId, { sequence, activeJobs, load, metadata = {} }) {
  return new WorkerHeartbeat({ worker_id: workerId, sequence, active_jobs: activeJobs, load, timestamp: new Date().toISOString(), metadata });
}
